### PR TITLE
acgan: Use Generator/Discriminator more closely resembling the ones from the referenced paper, and don't train discriminator to produce class labels for generated images

### PR DESCRIPTION
**Pull request by ozabluda**
This depends on PR https://github.com/fchollet/keras/pull/8452, and should be merged after it.

Generator:
1. Remove extra embedding layer.
2. Remove extra 1x1 convolutions. 
3. Use transposed convolutions instead of NN upsample+conv (=conv with fractional slides). Note that this corresponds to the paper, but contrary to author's own later advice in [Deconvolution and Checkerboard Artifacts](https://distill.pub/2016/deconv-checkerboard/). Still, I found that paper network works better on MNIST in this example.

Discriminator:
1. Use LeakyReLU slope 0.2, instead of default 0.3

These changes
1. made generated images more diverse (for example, thick and thin are on the same epoch, compare https://github.com/fchollet/keras/pull/8482#issuecomment-344401396 to https://github.com/fchollet/keras/pull/8409#issue-271565775)
2. Can tolerate harder soft_labels (all the way to 1.0 works without generator collapse, but the results are not as good as 0.9).
3. Reduced the number of trainable parameters in generator by 65% from 8,698,356 to 3,182,580.
4. Reduced training time per epoch by 50% from 90 sec to 45 sec on Titan X Maxwell.

### CODE REVIEW

**Comment by ozabluda** [comment]
Below, left image is generated, right image is actual MNIST dataset. In generated images, latent vector is the same per-row (see PR https://github.com/fchollet/keras/pull/8409)

Epoch 32:
plot_epoch_032_generated

Epoch 45:
plot_epoch_045_generated

Epoch 48:
plot_epoch_048_generated

**Comment by fchollet** [comment]
@lukedeo if your time allows, could you check out this PR as well?

**Comment by lukedeo** [comment]
@fchollet Overall, this looks good and is a welcome change.

@ozabluda you mention that you remove the extra embedding, but I don't see it removed here... the embedding wasn't mentioned in the paper, but I found the hadamard-y interaction to be useful. Not sure what the intention was here.

**Comment by ozabluda** [comment]
This extra "embedding" was removed:
https://github.com/fchollet/keras/pull/8482/files#diff-dc5bc2fdbf2d3f3268e301900b75e68eL52

Actual `Embedding`:
https://github.com/fchollet/keras/pull/8482/files#diff-dc5bc2fdbf2d3f3268e301900b75e68eR76

and its Hadamard product (which is cool) are untouched:
https://github.com/fchollet/keras/pull/8482/files#diff-dc5bc2fdbf2d3f3268e301900b75e68eR79

Please note that this PR is on top of PR https://github.com/fchollet/keras/pull/8452

**Comment by lukedeo** [comment]
Ah ok I see, sure, that extra hidden rep is extraneous and makes it more in-line with the paper.

Re: #8452, how noticeable is the improvement in quality when the discriminator receives zero weight for the auxiliary task on generated samples? If we're going to go full paper reproduction mode, my thought is to go all the way. I could be fairly easily convinced though, as having the discriminator minimize classification accuracy on generated samples can lead to the generator not examining sections of the support where the the likelihood ratio approaches a flat posterior

**Comment by ozabluda** [comment]
A brief history:

Recently acgan example was broken - generating all-black images. The minimal first fix for it was to introduce soft real/fake labels, PR "acgan: Fix generator producing pure black images" https://github.com/fchollet/keras/pull/8383, see generated examples there. This fix was either mandatory for later fixes (to prevent all-black), or was producing batter results. So far, all further fixes made soft labels harder and harder, which is an evidence that those are good fixes).


Second fix was PR "acgan: don't train discriminator to produce class labels for generated images" https://github.com/fchollet/keras/pull/8452, see generated examples there for how much more diverse they are than before. See https://github.com/fchollet/keras/pull/8452#issuecomment-343249641 for alternatives, and https://github.com/fchollet/keras/pull/8452#issuecomment-343705650 for the history how it ended up in the paper and the justification of its removal from the example (initially I was also hesitant to remove it, see the second sentence here: https://github.com/fchollet/keras/pull/8452#issue-272669339). It's better to keep the discussion of that PR in that PR.

Third fix is this PR. It depends on second and first.

**Comment by ozabluda** [comment]
Several PRs were merged into this PR. See https://github.com/fchollet/keras/pull/8409#issuecomment-346877248

File: examples/mnist_acgan.py [hunk 1]
```diff
@@ -185,6 +182,13 @@ def build_discriminator():
         num_batches = int(x_train.shape[0] / batch_size)
         progress_bar = Progbar(target=num_batches)
 
+        # don't train discriminator to produce class labels for generated
+        # images. To preserve total weight of the auxilary classifier,
+        # take real image samples with weight 2.
```
Comment (fchollet) [inline examples/mnist_acgan.py]: The meaning of this comment is not clear, please rephrase/clarify

File: examples/mnist_acgan.py [hunk 2]
```diff
@@ -185,6 +182,13 @@ def build_discriminator():
...
+        disc_sample_weight = [np.ones(2 * batch_size, dtype=np.float32),
```
Comment (fchollet) [inline examples/mnist_acgan.py]: This uses `float32` for one but not the other; it's not really necessary for either (memory is not an issue and they get cast anyway). In any case, better to be consistent

**Comment by ozabluda** [comment]
Review comments addressed.
